pkg/controller/drain/drain_controller.go

```diff
@@ -182,30 +181,25 @@ func (ctrl *Controller) logNode(node *corev1.Node, format string, args ...interf
 	klog.Infof("node %s: %s", node.Name, msg)
 }
 
-func (ctrl *Controller) handleNodeEvent(node interface{}) {
-	n := node.(*corev1.Node)
-	klog.V(4).Infof("Updating Node %s", n.Name)
-	ctrl.enqueueNode(n)
-}
+func (ctrl *Controller) handleNodeEvent(oldObj, newObj interface{}) {
 
-func (ctrl *Controller) enqueue(node *corev1.Node) {
-	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(node)
-	if err != nil {
-		utilruntime.HandleError(fmt.Errorf("Couldn't get key for object %#v: %v", node, err))
-		return
-	}
+	var oldNode, newNode *corev1.Node
 
-	ctrl.queue.Add(key)
-}
+	newNode = newObj.(*corev1.Node)
+	klog.V(4).Infof("Updating Node (drain controller) %s", newNode.Name)
 
-func (ctrl *Controller) enqueueRateLimited(node *corev1.Node) {
-	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(node)
-	if err != nil {
-		utilruntime.HandleError(fmt.Errorf("Couldn't get key for object %#v: %v", node, err))
+	if oldObj != nil {
+		oldNode = oldObj.(*corev1.Node)
+	} else {
+		ctrl.enqueueNode(newNode)
 		return
 	}
 
-	ctrl.queue.AddRateLimited(key)
+	// If the desiredDrain annotation are identical between oldNode and newNode, no new action is required by the drain controller
+	if oldNode.Annotations[daemonconsts.DesiredDrainerAnnotationKey] == newNode.Annotations[daemonconsts.DesiredDrainerAnnotationKey] {
```

Comment:
So my main concern is that this will cause long drains to be lost (but maybe it can't happen?) Could you help me double check this thinking process? Right now, how the drainers work is that:

1. if the current and desired drains are different, we kick off a drain process. There are 5 parallel worker threads in the controller so at most 5 nodes are being processed at a time.
2. the drain happens in 90s intervals, so if we don't finish (long drains, pdbs) we relinquish that drainer, and set ourselves back in the queue, which allows other nodes to drain

Now I think that process still works since that requeue's the node, but if the controller dies for some reason inbetween (e.g. if the node that the controller on gets drained), and a new controller starts up, this will always be true for a node that's halfway drained I think? So in theory nodes that didn't finish draining would be stuck there since it doesn't get queue'ed? Or is that case covered since `oldObj == nil` so it always gets enqueue'd above?

Given that CI is passing maybe this is fine? Probably best to check with some "heavier" workloads though during an upgrade process.

Reply:
>So in theory nodes that didn't finish draining would be stuck there since it doesn't get queue'ed? Or is that case covered since oldObj == nil so it always gets enqueue'd above?

Correct, since the cache gets filled up as the controller sstarts again, the very first AddNode() callback should attempt to requeue the node. 

> the drain happens in 90s intervals, so if we don't finish (long drains, pdbs) we relinquish that drainer, and set ourselves back in the queue, which allows other nodes to drain

I think I follow this too - just to double check my understanding though: This requeue here is via `enqueueAfter()` and not through `handleNodeEvent()` and hence is not guarded by the annotations check. So, we should be covered for the long running drain timeout/errors case.